arithmetization/src/main/java/net/consensys/linea/zktracer/module/ecdata/EcData.java

```diff
@@ -106,10 +106,53 @@ public void callEcData(
       case PRC_ECMUL -> ecMulEffectiveCall.addPrecompileLimit(1);
       case PRC_ECRECOVER -> ecRecoverEffectiveCall.addPrecompileLimit(1);
       case PRC_ECPAIRING -> {
-        // TODO: @Lorenzo @Olivier complete
-        //  ecPairingG2MembershipCalls.addPrecompileLimit();
-        //  ecPairingMillerLoops.addPrecompileLimit();
-        //  ecPairingFinalExponentiations.addPrecompileLimit();
+        // TODO: @Olivier @Lorenzo: review
+
+        // ecPairingG2MembershipCalls case
+        // NOTE: see EC_DATA specs Figure 3.5 for a graphical representation of this case analysis
+        if (!ecDataOperation.internalChecksPassed()) {
+          ecPairingG2MembershipCalls.addPrecompileLimit(0);
+          // The circuit is never invoked in the case of internal checks failing
+        }
+        // NOTE: the && of the conditions may seem not necessary since in the specs
+        // !internalChecksPassed => !notOnG2AccMax
+        // however, in EcDataOperation implementation the notOnG2AccMax takes into consideration
+        // only large points G2 membership
+        // , and it has to be && with internalChecksPassed to compute the actual
+        // NOT_ON_G2_ACC_MAX to trace
+        if (ecDataOperation.internalChecksPassed() && ecDataOperation.notOnG2AccMax()) {
+          ecPairingG2MembershipCalls.addPrecompileLimit(1);
+          // The circuit is invoked only once if there is at least one point predicted to be not on G2
+        }
+        if (ecDataOperation.internalChecksPassed()
+            && !ecDataOperation.notOnG2AccMax()
+            && ecDataOperation.overallTrivialPairing().getLast()) {
+          ecPairingG2MembershipCalls.addPrecompileLimit(0);
+          // The circuit is never invoked in the case of a trivial pairing
+        }
+        if (ecDataOperation.internalChecksPassed()
+            && !ecDataOperation.notOnG2AccMax()
+            && !ecDataOperation.overallTrivialPairing().getLast()) {
+          ecPairingG2MembershipCalls.addPrecompileLimit(
+              ecDataOperation.circuitSelectorG2MembershipCounter());
+          // The circuit is invoked as many times as there are points predicted to be on G2
+        }
+
+        final boolean predictedToBeValidSetOfPairings =
+            ecDataOperation.circuitSelectorEcPairingCounter() > 0;
+
+        // ecPairingMillerLoops case
+        // NOTE: the number of Miller Loops is equal to the number of pairings in a predicted to be
+        // valid set
+        // , 0 otherwise
+        ecPairingMillerLoops.addPrecompileLimit(
+            predictedToBeValidSetOfPairings ? ecDataOperation.totalPairings() : 0);
```

Comment:
The code defines `totalPairings` to be the `callDataSize / 192` but this may include trivial pairs. I.e. pairs where either (or both) points are the point at infinity. Are we sure such pairs are allowed in the gnark circuit ? I genuinely don't remember. If not we should filter such pairs out (given that we are in the success case.)

Reply:
Points at infinity are not sent to the circuit and do not activate the circuit selector / require extra computation.